poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	longPair *config.KeyPair, t, r int, insurers []abstract.Point) *Promise {
+	p.id = secretPair.Public.String()
+	p.t = t
+	p.r = r
+	p.n = len(insurers)
+	p.shareSuite = secretPair.Suite
+	p.keySuite = longPair.Suite
+	p.pubKey = longPair.Public
+	p.insurers = insurers
+	p.secrets = make([]abstract.Secret, p.n, p.n)
+
+	// Verify that t <= r <= n
+	if p.n < p.t {
+		panic("Not enough insurers for the secret")
+	}
+	if p.r < p.t {
+		p.r = p.t
+	}
+	if p.r > p.n {
+		p.r = p.n
+	}
+
+	// Create the public polynomial and private shares. The number of shares
+	// should be equal to the number of insurers.
+	pripoly := new(poly.PriPoly).Pick(p.shareSuite, p.t,
+		secretPair.Secret, random.Stream)
+	prishares := new(poly.PriShares).Split(pripoly, p.n)
+	p.pubPoly = poly.PubPoly{}
+	p.pubPoly.Commit(pripoly, nil)
+
+	// Populate the secrets array with the shares encrypted by a Diffie-
+	// Hellman shared secret between the promiser and appropriate insurer
+	for i := 0; i < p.n; i++ {
+		diffieBase := p.keySuite.Point().Mul(insurers[i], longPair.Secret)
+		p.secrets[i] = p.diffieHellmanEncrypt(prishares.Share(i), diffieBase)
+	}
+	return p
+}
+
+/* Initializes a Promise for unmarshalling
+ *
+ * Arguments
+ *    shareSuite = the suite used for shared secrets within the Promise
+ *    keySuite   = the suite used for long term keys within the Promise
+ *
+ * Returns
+ *   An initialized Promise ready to be unmarshalled
+ */
+func (p *Promise) UnmarshalInit(shareSuite, keySuite abstract.Suite) *Promise {
+	p.shareSuite = shareSuite
+	p.keySuite = keySuite
+	return p
+}
+
+/* Verifies that the Promise was constructed correctly.
+ *
+ * Arguments
+ *    promiserKey = the long term key the caller believes the Promise to be from
+ *
+ * Return
+ *   an error if the promise is malformed, nil otherwise.
+ *
+ * TODO Consider more ways to verify (such as making sure there are no duplicate
+ *      keys in p.insurers or that the promiser's long term public key is not in
+ *      p.insurers).
+ */
+func (p *Promise) VerifyPromise(promiserKey abstract.Point) error {
+	// Verify t <= r <= n
+	if p.t > p.n || p.t > p.r || p.r > p.n {
+		return errors.New("Invalid t-of-n shares Promise. Expected: t <= r <= n")
+	}
+	if !promiserKey.Equal(p.pubKey) {
+		return errors.New("Long term public key of Promise differs from what is expected.")
+	}
+	// There should be a secret and public key for each of the n insurers.
+	if len(p.insurers) != p.n || len(p.secrets) != p.n {
+		return errors.New("Insurers and secrets array should be of length promise.n")
+	}
+	return nil
+}
+
+// Returns the id of the Promise
+func (p *Promise) GetId() string {
+	return p.id
+}
+
+/* Given a Diffie-Hellman shared secret, encrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to encrypt
+ *    diffieBase  = the DH shared secret
+ *
+ * Return
+ *   the encrypted secret
+ */
+func (p *Promise) diffieHellmanEncrypt(secret abstract.Secret,
```

Comment:
How about just having this function return the diffieSecret without doing an Add.  Then you can eliminate diffieHellmanDecrypt, use this one function for both encryption and decryption, and do the Add or Sub in the caller.